### ISSUE TITLE
[Training] [C# API] TrainingSession ToBuffer() function returns type which is unusable

### ISSUE THREAD

**Issue by peratrepic** [microsoft/onnxruntime#17584]
### Describe the issue

I am using C# OnnxRuntime.Training API. I am trying to pull the trained weights out of the TrainingSession, which is proving to be impossible since TrainingSession.ToBuffer(true) returns FixedBufferOnnxValue which is a class without any functions except constructors. How are we supposed to use this result value from ToBuffer(), am I missing something?

### To reproduce

How to do this:

```
FixedBufferOnnxValue trainedParamsBuffer = m_trainingSession.ToBuffer(true);

float[] outputBuffer = trainedParamsBuffer.???????
```

### Urgency

This is blocking me from implementing Federated Learning use case with ONNX runtime since I cannot get trained parameters from the model. Only other option I have is to save new checkpoint, and then to manually do the diff `checkpoint2 - checkpoint1` on the server.

### ONNX Runtime Installation

Released Package

### ONNX Runtime Version or Commit ID

1.15.1

### PyTorch Version

2.01

### Execution Provider

Default CPU

**Comment by pranavsharma** [comment]
@askhade @baijumeswani - can one of you take a look?

**Comment by baijumeswani** [comment]
@peratrepic this might be an error from my end. Apologies you had to go through the trouble to find the API not usable.

I am addressing this in this pull request (https://github.com/microsoft/onnxruntime/pull/17364).

As you have highlighted, the passing the checkpoint might be (an inconvenient) a workaround.

**Comment by peratrepic** [comment]
Hi @baijumeswani , no problem, thank you for the quick response.

In the meantime, I am trying to understand what is the correct flow that was envisioned for onnx on device training. There are two options available that I see:
- Save updated checkpoint on device, send it to service, manually extract data from checkpoint protobuf tensors, average data, manually create new checkpoint with updated data, send new checkpoint to devices in the next round, load updated parameters by constructing TrainingSession with the new checkpoint.
- Get updated parameters on device with TrainingSession.ToBuffer(), send parameters buffer to service, average data, send new parameters buffer to devices in the next round, load updated parameters with TrainingSession.FromBuffer().

The second option seems much better and seems like it is the expected way to do this, but beside the issue that you are currently fixing there is another issue. TrainingSession has to be constructed with CheckpointState, which means that we always have to send some CheckpointState to device, which is unnecessary waste of IO since we are also sending the updated parameters buffer and loading it with TrainingSession.FromBuffer(), overwriting whatever was loaded from the checkpoint. The API doesn't provide us a way to construct TrainingSession only from the model architecture files, and then fill the parameters from buffer.

Could you please share some insights about this, what is the right way we are supposed to use the API for on device training?

**Comment by baijumeswani** [comment]
@peratrepic

The second bullet point is the expected flow for federated learning with on-device training with ONNX Runtime.

In addition to having the parameter weights, the checkpoint file also contains some other important data that is needed to create the training session. Some of this data includes:

- Metadata on the parameters themselves like the parameter names, whether or not the parameter requires gradient computation.
- It may also include the runtime properties that the user may set such as epoch number, best loss seen so far and anything that the user may deem necessary to save in the checkpoint that can be used later on while resuming training.
- It may also include the optimizer state if the user requests that information to be saved.

As such, the checkpoint is a necessary starting point in order to create the training session.

Having said that, your point is a valid one. Essentially, we are duplicating the loading of the weights when creating the training session, and then loading the new weights on top of the training session. For now, we don't have a way to work around this duplication of work (unless the training session is kept alive in memory).

We did have some discussion internally to address this. One idea that was floated was to create a minimal checkpoint file (that only contains the metadata needed to create the training session) and then use the FromBuffer method to load the parameters into the training session. Now, the problem with this approach is that we will need two sets of checkpoint files because at least one of the two (device or server) must hold the real parameter weights for ToBuffer and FromBuffer to work well.

While this work has not been prioritized yet, it is something we are aware of and will look to fix in the future.

In the meantime, depending on the scenario, you could consider just stashing the old checkpoint file and packaging it as a part of your application on the device (to save on network IO), or use the approach in the first bullet point you highlighted.

**Comment by peratrepic** [comment]
Thank you so much @baijumeswani, this was all very helpful info!

> you could consider just stashing the old checkpoint file

Good idea, this will work in my scenario!

**Comment by peratrepic** [comment]
Hi @baijumeswani, it seems like this fix wasn't included in the 1.16.1 release, do you maybe have some estimate when is it going to be included into official release?

I've tried using nightly version of ONNX runtime to test it out, but I get this error when I try to load checkpoint from file:
```
CheckpointState checkpointState = CheckpointState.LoadCheckpoint(CHECKPOINT_PATH);

>>> '[ErrorCode:Fail] open file checkpoint fail, errcode = 5 - Access is denied.'
```
It's a very weird error because it doesn't show with official release versions, do I need to give some special access permissions in order for prerelease versions to work?

**Comment by baijumeswani** [comment]
> Hi @baijumeswani, it seems like this fix wasn't included in the 1.16.1 release, do you maybe have some estimate when is it going to be included into official release?

It will be released in ONNX Runtime 1.16.2 release. Scheduled to be out in the first/second week of November.

```
CheckpointState checkpointState = CheckpointState.LoadCheckpoint(CHECKPOINT_PATH);

>>> '[ErrorCode:Fail] open file checkpoint fail, errcode = 5 - Access is denied.'
```

This does seem strange. I am not sure why this is the case. I can try it on my end to see if I can reproduce the error.

**Comment by GeorgeS2019** [comment]
Please add the label for csharp
https://github.com/microsoft/onnxruntime/labels/api%3ACSharp

**Comment by peratrepic** [comment]
Thank you @baijumeswani , no need to investigate this error, it turned out that in version 1.16 of `onnxruntime-training` python sdk it was changed that checkpoint is no longer generated into `checkpoint\paramtrain_tensors.pbseq`, but is now generated into `checkpoint` file with no extension in the same folder with other model artifacts. And the nightly version of the C# Microsoft.ML.OnnxRuntime.Training sdk apparently expects such structure in the model artifacts folder, and gives this error with the old folder structure.

**Comment by baijumeswani** [comment]
> The API doesn't provide us a way to construct TrainingSession only from the model architecture files, and then fill the parameters from buffer.

We now have added a new checkpoint (called the nominal checkpoint) which would allow the user to generate a minimal checkpoint for the device and a complete checkpoint for the server. Please look at https://github.com/microsoft/onnxruntime/pull/19232 for more details.
